### DISCUSSION TITLE
antsRegistration / antsApplyTransforms rounding error?

### DISCUSSION THREAD

**Discussion by aplumley** [ANTsX/ANTs#1439] [Q&A] [answered]
Hello,

I am trying to use antsApplyTransforms to apply the transformations output from 15 iterations of antsRegistration (using antsMultivariateTemplateConstruction2) to brain images. There are 30 transformations in total (15 GenericAffine and 15 Warps). But the result is incorrect - the output from antsApplyTransforms is incorrectly transformed (no longer brain-shaped). 

Is there much rounding involved when the transforms are saved? I am wondering whether the cumulation of rounding error may be causing some issues. 

Thanks in advance!

**Comment by cookpa** [comment, marked as answer]
I don't think there's any rounding, which is why the warp files tend to be so large even when compressed. 

I'm not sure how you are applying the warps but it sounds like something unusual. Each iteration (-i) of `antsMultivariateTemplateConstruction2.sh` should produce new warps, they are not designed to be applied in sequence.

**Reply by aplumley** [reply]
Thanks, this was useful. I was applying them in sequence. So the Affine and Warp files from the last iteration should be the one which was used to form the final output template, I guess.

**Reply by cookpa** [reply]
Given an initial template (or the template from the previous iteration), each iteration does the following

1. Register each moving image to the template
2. Average the deformed moving images in the template space
3. Apply sharpening to deblur the average image
4. Compute the average moving transform, and move the template closer to an unbiased shape and position. This step is a bit complicated but the effect is to minimize the "distance" between the template and the input images, so that the template remains representative and unbiased. 

This paper has more details on the [theory of template construction](https://pubmed.ncbi.nlm.nih.gov/19818860/)

So, after the last iteration, the final template has had step 4 applied to it. The warps left in the output directory are the registration of the moving images to the previous template (you can find it in the `intermediateTemplates/` directory). If the process has converged, then this template should be quite close to the final one.

**Reply by aplumley** [reply]
Thanks very much for the information. For context, I am trying to transform "image y" in the same way that "image x" was transformed with the templateReconstruction script, which included 15 iterations. So I just want to apply the same transformations that image x underwent. Is that possible? I am a little confused -from looking at the metriclog files at each iteration, it appears that this is the process:

N4BiasFieldCorrection 
-i [original image x] 
-o [image x]_Repaired

antsApplyTransforms 
-i [image x]_Repaired
-r [previous_iteration*]_reconstructedTemplate0       (* or the initial template)
-t [this_iteration]_Warp [this_iteration]_GenericAffine 
-o [this_iteration]_WarpedToTemplate

But this approach did not work when I implimented it with image y. I do not understand why N4BiasFieldCorrection is done at every iteration, since it appears to always take the original image x as input, and does not appear to have any input relating to the template. Am I missing something?

**Reply by cookpa** [reply]
I believe N4 is done at each iteration because it was easier to code, and with parallel processing, it didn't add too much extra time to a SyN iteration. 

These days, I always do N4 ahead of time along with other preprocessing and QC steps. This not only saves time, but lets me see exactly what is being used to create the template average.

In your metric logs, you should see N4, antsRegistration, then antsApplyTransforms. The transform is unique to every input image, and is named in the logs. The transform for image x is not the same as for image y. If image y was not in the template input set, you need to use `antsRegistration` to get the transform. See the scripts `antsRegistrationSyN.sh` and `antsRegistrationSyNQuick.sh` for a simplified starting point.

**Reply by gdevenyi** [reply]
You may also be interested in my alternative re-implementation of the ants template algorithm,
https://github.com/CoBrALab/optimized_antsMultivariateTemplateConstruction

**Comment by tangYING2022** [comment]
Hi, l meet two problems when l'm trying to build a template.

- 1）the script l use is `antsMultivariateTemplateConstruction2.sh`, and the iteration was set 4 runs. However, when it started the forth run, the shell in ubuntu didn't continue to run job*.sh and no files of job*metriclog were generated. Why did this happen? And how can l deal with it?

- 2）l used to see a post about doing iteration separately (https://sourceforge.net/p/advants/discussion/840260/thread/23bc83c087/.)  But the result of my forth run is not as good as the third template. Is it resonable?
_Here is the command：
antsMultivariateTemplateConstruction2.sh -d 3 -o T1 -a 1 -c 1 -e 1 -g 0.25 -i 1 -z initial_template -k 1 -w 1 -q 100x100x70x20 -f 6x4x2x1 -s 3x2x1x0 -n 0 -r 1 -l 1 -m CC -t SyN -y 1 *.nii.gz_

l 'm looking forward to your reply.^v^
